---
title: "Figure 4: Drugs & 7 donors"
---
# Please be aware that this is a reserach code and far from perfect ;P Cheers!

# Setup
```{r setup, echo =FALSE}
.libPaths('C:\\Users\\zuendj\\Documents\\R4Libs')
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "P:/Shared_documents/Serafina Pl√ºss/20230320_sequencing_results/Janina", cho = TRUE, warning = FALSE, message = FALSE)
 # devtools::install_github("tidyverse/tidyverse")
library(tidyverse)
library(openxlsx)
library(ggh4x)
library(microViz)
library(ggpmisc)
library(ggpubr)
library(rstatix)
```

```{r, echo =FALSE}
# source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_taxa_tests.R") 
# source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_normalisation.R") 
# source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_alpha.R") 
# source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_beta.R") 
# source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_heatmap.R")
# source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_taxa_tests.R")
# source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_varia.R")
# source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_beta.R")

#get metadata file: load this function
sample.data.frame <- function(ps) {
  return(as(phyloseq::sample_data(ps), "data.frame"))}

facet_labeller <- function(variable,value){
  return(names[value])}
```

# Get the phyloseq object 
```{r}
HTP <-readRDS("P:/Shared_documents/Former members/Serafina Pl¸ss/13_High-throughput_JANINA_part2/03_sequencing_data/dada2/phyloseq.RDS")
```


```{r}
#fix tax table
HTP %>% 
  phyloseq_get_strains()-> HTP

library(microViz)

noNA = !is.na(tax_table(HTP)[,"Genus"]) & !is.na(tax_table(HTP)[,"Species"])
tax_table(HTP)[noNA][,"Species"] = paste(tax_table(HTP)[noNA][,"Genus"], tax_table(HTP)[noNA][,"Species"])

HTP%>%
  tax_fix()->HTP

tax<-as.data.frame(HTP@tax_table)

tax$Genus<-gsub("_7","", tax$Genus)
tax$Species<-gsub("_7","", tax$Species)
tax$Genus<-gsub("_9","", tax$Genus)
tax$Species<-gsub("_9","", tax$Species)

tax%>%
  subset(Family %in% c("Oscillospiraceae"))

tax$Genus<-gsub("Incertae Sedis","Ruminococcaceae", tax$Genus)
tax$Species<-gsub("Incertae Sedis","Ruminococcaceae", tax$Species)
tax$Species<-gsub("Candidatus Soleaferrea","Ruminococcaceae", tax$Species)
tax$Genus<-gsub("Candidatus Soleaferrea","Ruminococcaceae", tax$Genus)

tax_table(HTP)<-as.matrix(tax, row_names =T)
```

### Get and update metadata file
```{r, echo=FALSE}
HTP%>%
  physeq_add_metadata(physeq = .,
                      metadata = "P:/Shared_documents/Former members/Serafina Pl¸ss/13_High-throughput_JANINA_part2/03_sequencing_data/mapping_file_updated.xlsx" %>%
                        readxl::read_xlsx(),
                      sample_column = "sample_name") -> HTP

meta_htp<-sample.data.frame(HTP)
meta_htp$treatment <- recode_factor(meta_htp$treatment, "control" = "a_control")
meta_htp$treatment <- recode_factor(meta_htp$treatment, "5-FU" = "b_5-FU")
meta_htp$donor <- recode(meta_htp$donor, "FDS" = "WPL")
meta_htp$treatment<-as.character(meta_htp$treatment)
sample_data(HTP)<-meta_htp
```

```{r}
ps_strain_filt <- HTP

###filter 1
# threshold in %
threshold = 0.1

# filter per sample to remove all ASV lower in abundance than threshold (%)
otu_table(ps_strain_filt) <- otu_table(ps_strain_filt) %>%
  as.data.frame() %>%
  dplyr:: mutate(across(everything(), ~ ifelse(. <= sum(.) * (threshold/100), 0, .))) %>% otu_table(., taxa_are_rows = TRUE)

# remove all ASVs that are not present in any sample
HTP <- ps_strain_filt %>%
  filter_taxa(function(x) sum(x > 0) > 0, TRUE)

HTP%>%
  subset_samples(experiment %in% "HTP")%>%
  subset_samples(!donor %in% c("WDR"))%>% #exclude because of low diversity
  subset_samples(treatment %in% c("a_control" , "DMSO" , "omeprazole","ciprofloxacin", "b_5-FU"))->phylo_drugs
```

```{r, echo=FALSE}
#BiocManager::install("microbiome/mia")
library(mia)
library(miaViz)
library(ALDEx2)
```

```{r}
drugs <- makeTreeSummarizedExperimentFromPhyloseq(phylo_drugs) 
```


```{r}
tse_drugs <- drugs %>% agglomerateByRank(rank = "Family")
```

###Aldex: NO Prevalence filter-> filter later with abundance
```{r}
substrates <-c("DMSO", "ciprofloxacin","omeprazole", "b_5-FU")
all_drugs <-data.frame(matrix(ncol=14, nrow=0))

for (i in (1:length(substrates))){
  tse_substrates <-tse_drugs[ , tse_drugs$treatment %in% c(substrates[i], "a_control")]
      # tse <- subsetByPrevalentTaxa(tse_substrates, detection = 0, prevalence=0.5, include_lowest= TRUE)
      result_name <- paste(substrates[i])
      print(result_name)
      count <- as.vector(tse_substrates$treatment)
      x <- aldex.clr(
      reads = assay(tse_substrates),
     tse_substrates$treatment, 
      mc.samples = 1000)
    # calculates expected values of the Welch's t-test and Wilcoxon rank test on
    # the data returned by aldex.clr
      x_tt <- aldex.ttest(
        x, 
        paired.test = FALSE, 
        verbose = FALSE)
      # determines the median clr abundance of the feature in all samples and in
      # groups, the median difference between the two groups, the median variation
      # within each group and the effect size, which is the median of the ratio
      # of the between group difference and the larger of the variance within groups
      x_effect <- aldex.effect(x, CI = TRUE, verbose = FALSE)
      # combine all outputs 
      aldex_out <- data.frame(x_tt, x_effect)

      result <-aldex_out
      result[,14] <- result_name
      print(colnames(result))
      
      names(result)[6]<-paste0("rab.win.drug")
      names(result)[7]<-paste0("rab.win.H2O")
      result$Family<-row.names(result)
      
      all_drugs <- rbind(all_drugs, result)
          }

all_drugs %>%
  filter(wi.eBH < 0.05)->out
```

```{r}
unloadNamespace("miaViz")
unloadNamespace("mia")

phylo_drugs%>%
  physeq_glom_rename(taxrank = "Family", 
                     speedyseq = TRUE)%>%
  subset_samples(!donor %in% "WDR") %>%
   microbiome::transform("clr") %>% 
   psmelt() ->df_abund

clr_drug<-left_join(df_abund, as.data.frame(phylo_drugs@tax_table))
mean_abund<-setNames(aggregate(clr_drug$Abundance, by=list(clr_drug$Family), FUN = "mean"), c("Family", "abund"))
clr_drug<-left_join(clr_drug, mean_abund)
mean_abund$Family<-as.factor(mean_abund$Family)

clr_drug%>%
  mutate(Family = fct_reorder(Family, clr_drug$abund)) ->hoi #ran out of good df names

ordered<-unique(as.character(hoi$Family))[1:15]

clr_drug%>%
  subset(Family %in% c(ordered))->clr_drug

clr_diff<-setNames(as.data.frame(matrix(0,0,4)), c("diff", "Family", "treatment", "donor"))

drugs_cond<- c( "ciprofloxacin", "b_5-FU", "omeprazole")
drugs_comparison<- c( "a_control", "a_control", "DMSO")
donor_cond<-unique(clr_drug$donor)
donor_fam<-unique(clr_drug$Family)
i=1
for (m in 1: length(drugs_cond)){
  clr_drug%>%
    subset(treatment %in% drugs_cond[m])->sub1
  
  for (j in 1:length(donor_cond)){
    sub1%>%
    subset(donor %in% donor_cond[j])->sub2
    
    for (k in 1:length(donor_fam)){
    sub2%>%
    subset(Family %in% donor_fam[k])->sub3
  
      if (nrow(sub3) > 0){
      
      clr_drug%>%
        subset(treatment %in% drugs_comparison[m] & donor %in% donor_cond[j] & Family %in% donor_fam[k] )->hoi
        
      clr_diff[i, c("diff")] <- sub3$Abundance - hoi$Abundance 
   
      clr_diff[i, c("Family")] <-donor_fam[k]
      clr_diff[i, c("treatment")] <-drugs_cond[m]
      clr_diff[i, c("donor")] <-donor_cond[j]
     
  
  i=i+1}
    }}}

clr_diff<-left_join(clr_diff, mean_abund)


clr_diff%>%
  mutate(Family = fct_reorder(Family, clr_diff$abund))%>%
  ggplot(aes(x=diff, y=Family))+geom_vline(xintercept = 0, linetype = "dashed", colour = "grey", size=1)+
   geom_boxplot(aes(fill=treatment), outlier.shape = NA)+
  geom_jitter(aes(group=treatment, colour=donor), position = position_jitterdodge(0.1) ,size=2, alpha=0.75)+
   scale_fill_manual(values = c( "cyan4" , "darkgoldenrod1", "#F8766D"),
                        labels = c(  "5-FU","Ciprofloxacin", "Omeprazole"), 
                       name="Drug")+
   theme(axis.text.y= element_text(size = 20, face="italic"), 
              axis.title.y = element_text(size=20), 
              axis.title.x = element_text(size=20),
              axis.text.x = element_text(size=20),
              legend.text = element_text(size = 20),
              legend.title = element_text(size= 20),
              plot.title = element_text(size=20), panel.background = element_rect(fill="white"),
        panel.border=element_rect(fill=NA))+
      ylab("")+xlab("Differences in clr-abundance")+labs(colour="Donor")+
      ggtitle("Drug-specifically altered families")+
  xlim(c(-10, 10))

ggsave(filename = "drugs_bars.jpeg",
       path = "C:/Users/zuendj/Desktop/03_data R/wp1/output/final",
       width = 27,
       height = 30,
       units = c("cm"))
```
# Aldex: NO Prevalence filter-> filter later with abundance
```{r}
substrates <-c("omeprazole")

all_drugs <-data.frame(matrix(ncol=14, nrow=0))

for (i in (1:length(substrates))){
  tse_substrates <-tse_drugs[ , tse_drugs$treatment %in% c(substrates[i], "DMSO")]
  
  
      tse <- subsetByPrevalentTaxa(tse_substrates, detection = 0, prevalence=0.5, include_lowest= TRUE)
      
      result_name <- paste(substrates[i])
      print(result_name)
      
      count <- as.vector(tse$treatment)
     
     
     
      x <- aldex.clr(
      reads = assay(tse),
     tse$treatment)
    # calculates expected values of the Welch's t-test and Wilcoxon rank test on
    # the data returned by aldex.clr
      x_tt <- aldex.ttest(
        x, 
        paired.test = FALSE, 
        verbose = FALSE)
      # determines the median clr abundance of the feature in all samples and in
      # groups, the median difference between the two groups, the median variation
      # within each group and the effect size, which is the median of the ratio
      # of the between group difference and the larger of the variance within groups
      x_effect <- aldex.effect(x, CI = TRUE, verbose = FALSE)
      # combine all outputs 
      
      aldex_out <- data.frame(x_tt, x_effect)
      
      result <-aldex_out
      result[,14] <- result_name
      print(colnames(result))
      
      names(result)[6]<-paste0("rab.win.drug")
      names(result)[7]<-paste0("rab.win.H2O")
      result$Family<-row.names(result)
      all_drugs <- rbind(all_drugs, result)
      }

all_drugs %>%
  filter(we.ep < 0.05)->out
```
  