---
title: "C-sources & 8 donors"
---

Link to Florentin's github page: https://htmlpreview.github.io/?https://github.com/fconstancias/DivComAnalyses/blob/master/SOP/community_analyses/Microbiome_R_tutorial.html#alpha-diversity

1. Getting ready: load all packages
-newest version of R
-rtools required 


```{r setup, echo =FALSE}
.libPaths('C:\\Users\\zuendj\\Documents\\R4Libs')
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "P:/Shared_documents/Serafina Pl√ºss/20230320_sequencing_results/Janina", cho = TRUE, warning = FALSE, message = FALSE)



# install.packages("vctrs")
# install.packages("vegan")
# install.packages("cli")

# library(vctrs)
# library(vegan)
# install.packages("ggplot2")
#library(ggplot2)
#install.packages('tinytex')
#install.packages('rmarkdown')
#install.packages("rlang")
#install.packages("nlme")
#install.packages("mgcv")
#install.packages("ape")
#library(nlme)
# install.packages("remotes")
# remotes::install_github("mikemc/speedyseq")
# install.packages("BiocManager")
# BiocManager::install("DESeq2")
#install.packages("fs")
# install.packages('devtools')
#install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
#unload("mia")
# install.packages("GUniFrac")

# install.packages(
#   "microViz",
#   repos = c(davidbarnett = "https://david-barnett.r-universe.dev", getOption("repos"))
# )
# library(microViz)

# devtools::install_github("tidyverse/tidyverse")
# library(tidyverse)
# library(openxlsx)
# # install.packages("ggalluvial")
# library(ggalluvial)
# 
# library(GUniFrac)
# library(rstatix)
# library(tinytex)
# library(rmarkdown)
# library(rlang)
# library(tidyverse)
# library(dplyr)
# library(mgcv)
# library(ape)
 # library(phyloseq)
# library(speedyseq)
# library(DESeq2)
# library(plyr)
# library(fs)
# library(devtools)
# library(pairwiseAdonis)
# library(ggpubr)

# install.packages("dplyr")
# library(metagMisc)

# devtools::install_github("vmikk/metagMisc")
```

```{r, echo =FALSE}
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_taxa_tests.R") 
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_normalisation.R") 
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_alpha.R") 
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_beta.R") 
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_heatmap.R")
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_taxa_tests.R")
source("https://raw.githubusercontent.com/fconstancias/metabaRpipe-source/master/Rscripts/functions.R")
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_varia.R")
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_beta.R")

#get metadata file: load this function
sample.data.frame <- function(ps) {
  return(as(phyloseq::sample_data(ps), "data.frame"))}
```

## 1. Get the phyloseq object 
```{r}
# unload("mia")
phyloseq <-readRDS("P:/Shared_documents/Former members/Serafina Pl¸ss/20230320_sequencing_results/Janina/dada2/phyloseq_phylo/phyloseq_phylo.RDS")

phyloseq%>%
  phyloseq_get_strains()->phyloseq
```


### Get and update metadata file
```{r, echo=FALSE}

phyloseq%>%
  physeq_add_metadata(physeq = .,
                      metadata = "P:/Shared_documents/Former members/Serafina Pl¸ss/20230320_sequencing_results/Janina/metadata_file_janina.xlsx" %>%
                        readxl::read_xlsx(),
                      sample_column = "sample_name") -> phyloseq


meta<-sample.data.frame(phyloseq)



ps_strain_filt <- phyloseq

###filter 1
# threshold in %
threshold = 0.1

# filter per sample to remove all ASV lower in abundance than threshold (%)
otu_table(ps_strain_filt) <- otu_table(ps_strain_filt) %>%
  as.data.frame() %>%
  dplyr:: mutate(across(everything(), ~ ifelse(. <= sum(.) * (threshold/100), 0, .))) %>% otu_table(., taxa_are_rows = TRUE)

# remove all ASVs that are not present in any sample
ps_strain_filt <- ps_strain_filt %>%
  filter_taxa(function(x) sum(x > 0) > 0, TRUE)


```

```{r, echo=FALSE}
#BiocManager::install("microbiome/mia")


library(mia)
library(miaViz)
library(ALDEx2)
```

```{r}
ps_strain_filt%>%
  subset_samples(experiment %in% "high-throughput")%>%
  # subset_taxa(Family %in% "Enterobacteriaceae")%>%
  subset_samples(condition %in% c("feces", "7C",
                                                 "3C+Muc",
                                                 "7C-Muc" ,
                                                 "3C" ,
                                                 "SS" ,
                                                 "Nut" ,
                                                 "Glc" ,
                                                 "H2O"  
                                                 ))->C

enetros <- makeTreeSummarizedExperimentFromPhyloseq(C) 

```


```{r}
tse_C <- enetros %>% agglomerateByRank(rank = "Family")
```

###Aldex: NO Prevalence filter-> filter later with abundance
```{r}

substrates <-c("7C",
                                                 "3C+Muc",
                                                 "7C-Muc" ,
                                                 "3C" ,
                                                 "SS" ,
                                                 "Nut" ,
                                                 "Glc" ,
                                                 "H2O" )


all_C <-data.frame(matrix(ncol=14, nrow=0))

for (i in (1:length(substrates))){
  tse_substrates <-tse_C[ , tse_C$condition %in% c(substrates[i], "feces")]
  
  
    
      
      result_name <- paste(substrates[i])
      print(result_name)
      
      count <- as.vector(tse_substrates$condition)
     
     
     
      x <- aldex.clr(
      reads = assay(tse_substrates),
     tse_substrates$condition)
    # calculates expected values of the Welch's t-test and Wilcoxon rank test on
    # the data returned by aldex.clr
      x_tt <- aldex.ttest(
        x, 
        paired.test = FALSE, 
        verbose = FALSE)
      # determines the median clr abundance of the feature in all samples and in
      # groups, the median difference between the two groups, the median variation
      # within each group and the effect size, which is the median of the ratio
      # of the between group difference and the larger of the variance within groups
      x_effect <- aldex.effect(x, CI = TRUE, verbose = FALSE)
      # combine all outputs 
      
      aldex_out <- data.frame(x_tt, x_effect)
      
      
      
      result <-aldex_out
      result[,14] <- result_name
print(colnames(result))
      
      names(result)[6]<-paste0("rab.win.feces")
      names(result)[7]<-paste0("rab.win.cultures")
      result$Family<-row.names(result)
      
      all_C <- rbind(all_C, result)
      
     
      
      }

all_C %>%
  subset(Family %in% "Family:Enterobacteriaceae")%>%
  filter(we.ep < 0.05)->out


out[,c(1,8,10, 14,15)]

BACKUP<-all

out %>%
  subset(Family %in% "Family:Enterobacteriaceae") %>%
  subset(wi.eBH < 0.05)
```

```{r}
unloadNamespace("miaViz")
unloadNamespace("mia")

```


```{r}

# library("dplyr")
C%>%
  physeq_glom_rename(taxrank = "Family", 
                     speedyseq = TRUE)%>%
   microbiome::transform("clr") %>% 
   psmelt() ->df_abund

clr_c<-left_join(df_abund, as.data.frame(C@tax_table))

mean_abund<-setNames(aggregate(clr_c$Abundance, by=list(clr_c$Family), FUN = "mean"), c("Family", "abund"))

clr_c<-left_join(clr_c, mean_abund)

mean_abund$Family<-as.factor(mean_abund$Family)


clr_c%>%
  subset(Family %in% c("Enterobacteriaceae"))->clr_entero

clr_diff<-setNames(as.data.frame(matrix(0,0,4)), c("diff", "Family", "condition", "donor"))

drugs_cond<- c( "7C",
                                                 "3C+Muc",
                                                 "7C-Muc" ,
                                                 "3C" ,
                                                 
                                                 "Glc" ,
                                                 "H2O" )
donor_cond<-unique(clr_entero$donor)
donor_fam<-unique(clr_entero$Family)

i=1
for (m in 1: length(drugs_cond)){
  clr_entero%>%
    subset(condition %in% drugs_cond[m])->sub1
  
  for (j in 1:length(donor_cond)){
    sub1%>%
    subset(donor_name %in% donor_cond[j])->sub2
    
    for (k in 1:length(donor_fam)){
    sub2%>%
    subset(Family %in% donor_fam[k])->sub3
  
      if (nrow(sub3) > 0){
      
      clr_entero%>%
        subset(condition %in% "feces" & donor_name %in% donor_cond[j] & Family %in% donor_fam[k] )->hoi
        
      clr_diff[i, c("diff")] <- sub3$Abundance[1] - hoi$Abundance [1]
   
      clr_diff[i, c("Family")] <-donor_fam[k]
      clr_diff[i, c("condition")] <-drugs_cond[m]
      clr_diff[i, c("donor")] <-donor_cond[j]
     
  
  i=i+1}
    }}}

clr_diff<-left_join(clr_diff, mean_abund)

clr_diff$condition<-factor(clr_diff$condition, levels=c("7C", "3C+Muc", "7C-Muc", "3C", "Glc", "H2O"))

clr_diff%>%
  # mutate(Family = fct_reorder(Family, clr_diff$abund))%>%
  ggplot(aes(x=diff, y=condition))+
  geom_vline(xintercept = 0, linetype = "dashed", colour = "grey", size=1)+
   geom_boxplot( outlier.shape = NA)+
  geom_jitter(aes(colour=donor), position = position_jitterdodge(0.01) ,size=3, alpha=0.75)+
   # scale_fill_manual(values = c( "cyan4" , "darkgoldenrod1"),
   #                      labels = c(  "5-FU","Ciprofloxacin"), 
   #                     name="Drug")+
   theme(axis.text.y= element_text(size = 20), 
              axis.title.y = element_text(size=20), 
              axis.title.x = element_text(size=20),
              axis.text.x = element_text(size=20),
              legend.text = element_text(size = 20),
              legend.title = element_text(size= 20),
              plot.title = element_text(size=20), panel.background = element_rect(fill="white"),
        panel.border=element_rect(fill=NA))+
      ylab("")+xlab("Differences in clr-abundances")+labs(colour="Donor")+
      ggtitle(expression(paste(italic("Enterobacteriaceae"), " levels")))+
  xlim(c(0, 8))+
  scale_y_discrete(labels = c('feces' = 'Feces', '7C' = '6C+Muc', '3C+Muc' = '3C+Muc', "Nut"="Nut", "SS"="SS", "7C-Muc"="6C", "3C"="3C", "H2O"="C-depleted", "Glc"="Glucose"))->p

show(p + theme(legend.position = "none"))
  # stat_pvalue_manual(stat.test,  label = "p.adj.signif", label.size = 8)

ggsave(filename = "c_sources_bars.jpeg",
       path = "C:/Users/zuendj/Desktop/03_data R/wp1/output/final",
       width = 14,
       height = 14,
       units = c("cm"))


51.7/14*27
65.3/17*35
```

```{r}
aggregate(clr_diff$diff, by=list(clr_diff$condition), FUN="median")
```

